ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/CommandStore.java

```diff
@@ -46,166 +46,119 @@
-    // is performed to see if anything new has been written (which would be signalled by the end
-    // offsets having changed). If something new has been written, the outer loop is repeated; if
-    // not, we're caught up to the end offsets we just polled and can continue.
-    do {
-      while (!offsetsCaughtUp(currentOffsets, endOffsets)) {
-        log.debug("Polling for prior command records");
+    /**
+     * Close the store, rendering it unable to read or write commands
+     */
+    @Override
+    public void close() {
+        closed.set(true);
+        commandConsumer.wakeup();
+        commandProducer.close();
+    }
+
+    /**
+     * Write the given statement to the command topic, to be read by all nodes in the current cluster.
+     * Does not return until the statement has been successfully written, or an exception is thrown.
+     * @param statementString The string of the statement to be distributed
+     * @param statement The statement to be distributed
+     * @param streamsProperties Any command-specific Streams properties to use.
+     * @return The ID assigned to the statement
+     * @throws Exception TODO: Refine this
+     */
+    public CommandId distributeStatement(
+            String statementString,
+            Statement statement,
+            Map<String, Object> streamsProperties
+    ) throws Exception {
+        CommandId commandId = commandIdAssigner.getCommandId(statement);
+        Command command = new Command(statementString, streamsProperties);
+        commandProducer.send(new ProducerRecord<>(commandTopic, commandId, command)).get();
+        return commandId;
+    }
+
+    /**
+     * Poll for new commands, blocking until at least one is available.
+     * @return The commands that have been polled from the command topic
+     */
+    public ConsumerRecords<CommandId, Command> getNewCommands() {
+        return commandConsumer.poll(Long.MAX_VALUE);
+    }
+
+    /**
+     * Collect all commands that have been written to the command topic, starting at the earliest
+     * offset and proceeding until it appears that all have been returned.
+     * @return The commands that have been read from the command topic
+     */
+    public List<Pair<CommandId, Command>> getPriorCommands() {
+        List<Pair<CommandId, Command>> result = new ArrayList<>();
+        for (ConsumerRecord<CommandId, Command> commandRecord : getAllPriorCommandRecords()) {
+            CommandId commandId = commandRecord.key();
+            Command command = commandRecord.value();
+            if (command != null) {
+                result.add(new Pair<>(commandId, command));
+            }
+        }
+
+        return result;
+    }
+
+    private List<ConsumerRecord<CommandId, Command>> getAllPriorCommandRecords() {
+        Collection<TopicPartition> commandTopicPartitions = getTopicPartitionsForTopic(commandTopic);
+
+        commandConsumer.poll(0);
```

Comment:
Let's also keep the comment for this poll:

// Have to poll to make sure subscription has taken effect (subscribe() is lazy)